### ISSUE TITLE
Add Check Support for Java 21 Record Pattern : New Check PatternVariableAssignment

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#14949]
child of #14890 

pattern variable assignment is considered bad programming practice. The pattern variable is meant to be a direct reference to the object being matched. Reassigning it can break this connection and mislead readers. we should violate mutating variables that are intended to represent matched patterns. So, a new check named PatternVariableAssignment

**Config:**

```XML
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="PatternVariableAssignment"/>
    </module>
</module>
```
---
**Example**
```Java
public class PatternVariables {
    public void testAssignment(Object obj) {
        if (obj instanceof String s) {
            s = "hello";             // violation
            System.out.println(s);
        }
        if (obj instanceof Rectangle(ColoredPoint x, ColoredPoint y)) {
            x = new ColoredPoint(1, 2, "red");   // violation
        }
        if (obj instanceof Rectangle(ColoredPoint(int x1,int x2,String c), _)) {
            c = "red";   // violation
        }
    }
}
```

**Comment by nrmancuso** [comment]
@mahfouz72 please update the description with the actual violation message we will use, then I can approve this.

**Comment by mahfouz72** [comment]
@nrmancuso done

**Comment by SteLeo1602** [comment]
I'm on it

**Comment by romani** [comment]
@nrmancuso , @mahfouz72 , name of check should be changed to PatternVariable**Re**assignment.

see reasoning at https://github.com/checkstyle/checkstyle/pull/16946#discussion_r2134406709

**Comment by romani** [comment]
@rdiachenko , please consider renaming Check.

**Comment by mahfouz72** [comment]
@romani

fwiw, the parameter version of this check is named: ParameterAssignment

https://checkstyle.org/checks/coding/parameterassignment.html#ParameterAssignment
https://checkstyle.org/checks/coding/innerassignment.html#InnerAssignment

---

IDEA also uses the term "Assignment" (e.g., Assignment to method parameter), not "Reassignment".
https://www.jetbrains.com/help/inspectopedia/Java-Assignment-issues.html

https://www.jetbrains.com/help/inspectopedia/AssignmentToMethodParameter.html

The check flags the act of assigning a value to a pattern variable, regardless of whether it's a reassignment or the first assignment

---
     if (obj instanceof String s) {
            s = "hello";                       
        }

It is technically a reassignment, because the variable `s` has already been assigned a value by the pattern match. but, for convenience and consistency, I referred to it simply as an assignment, since there is no explicit 's = ...` in the written code before this line. The initial assignment is implicitly done by the pattern match, not by a written statement.

**Comment by romani** [comment]
>fwiw, the parameter version of this check is named: ParameterAssignment

yes, this is why we named it like this from the beginning.
But we do not need consistency here, better to use better name.


>The check flags the act of assigning a value to a pattern variable, regardless of whether it's a reassignment or the first assignment

assignment happens at ` if (obj instanceof String s) ` , `s = "hello";  ` is reassignment.

-----------

https://docs.oracle.com/en/java/javase/17/language/pattern-matching-instanceof.html

> A pattern is a combination of a predicate, or test, that can be applied to a target and a set of local variables, called pattern variables, that are **assigned** values extracted from the target only if the test is successful. 

PMD use reassignment: https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#avoidreassigningparameters

IDEA use both terms in descriptions, not a title:
![Image](https://github.com/user-attachments/assets/f3d9dac4-ed90-4c01-a1e1-9947d6ae62f3)

Sonar use both terms: https://rules.sonarsource.com/java/RSPEC-1226
>Method parameters, caught exceptions and foreach variables' initial values should not be ignored
While it is technically correct to **assign to parameters** from within method bodies, doing so before the parameter value is read is likely a bug. Instead, initial values of parameters, caught exceptions, and foreach parameters should be, if not treated as final, **then at least read before reassignment**.

**Comment by romani** [comment]
world is split on this, sound like both is ok.
We can use Assignment, just because it is easy to read/pronounce.

But check will catch first explicit in code assignment

**Comment by romani** [comment]
Merged at https://github.com/checkstyle/checkstyle/pull/16946
